src/lib/x509/x509path.cpp

```diff
@@ -209,12 +214,76 @@ PKIX::check_chain(const std::vector<X509_Certificate>& cert_path,
    return cert_status;
    }
 
+namespace {
+
+Certificate_Status_Code verify_ocsp_signing_cert(
+      const X509_Certificate& signing_cert,
+      const X509_Certificate& ca,
+      const std::vector<X509_Certificate>& extra_certs,
+      const std::vector<Certificate_Store*>& certstores,
+      std::chrono::system_clock::time_point ref_time,
+      const Path_Validation_Restrictions& restrictions)
+   {
+   // RFC 6960 4.2.2.2
+   //    [Applications] MUST reject the response if the certificate
+   //    required to validate the signature on the response does not
+   //    meet at least one of the following criteria:
+   //
+   //    1. Matches a local configuration of OCSP signing authority
+   //       for the certificate in question, or
+   if(restrictions.trusted_ocsp_responders()->certificate_known(signing_cert))
+      return Certificate_Status_Code::OK;
+
+   // RFC 6960 4.2.2.2
+   //
+   //    2. Is the certificate of the CA that issued the certificate
+   //       in question, or
+   if(signing_cert == ca)
+      return Certificate_Status_Code::OK;
+
+   // RFC 6960 4.2.2.2
+   //
+   //    3. Includes a value of id-kp-OCSPSigning in an extended key
+   //       usage extension and is issued by the CA that issued the
+   //       certificate in question as stated above.
+
+   // TODO: Implement OCSP revocation check of OCSP signer certificate
+   // Note: This needs special care to prevent endless loops on specifically
+   //       forged chains of OCSP responses referring to each other.
```

Comment:
I figure this is worth discussing in more detail and might go beyond the scope of this refactoring. Nevertheless, please bear with me:

OCSP responder certificates can again offer an OCSP responder in their AIA extension. Clients compliant with [RFC 6960 4.2.2.2.1](https://www.rfc-editor.org/rfc/rfc6960#section-4.2.2.2.1) (and [TR-02103 8.5.5](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Publikationen/TechnischeRichtlinien/TR02103/BSI-TR-02103.pdf) (in German) for that matter) are obliged to recursively validate OCSP responses.

Currently, Botan's API cannot really support this for applications where `check_ocsp_online()` is not an option -- i.e. that are forced to provide all relevant OCSP responses up-front. `x509_path_validate()` takes vectors of `X509_Certificate` and `OCSP::Response` objects. The contents of those vectors must be in "lock step", i.e. the i-th `OCSP::Response` refers to the i-th `X509_Certificate`. Hence, there's no way an application could insert OCSP responses for "higher order" revocation checks of OCSP responder certificate as they don't actually show up in the invocation of `x509_path_validate()` in the first place.

Assuming that these "higher order" OCSP responses are quite a corner case anyway, I guess it should be fine to just not support them for now.

Though, I believe the API shortcoming is something we might want to consider. Even without the context of "higher order OCSP responses".

Applications that cannot rely on `check_ocsp_online()` and use the X.509 validation independently from TLS currently need to duplicate a fair share of the certificate path-building code. They must figure out themselves which certificates and OCSP responses are needed for `x509_path_validate()` to succeed. When working on _Bdrive_, we used to maintain quite some machinery to make this work (even without the "higher order" OCSP responses).

It would be great if the X.509 path validation would provide hooks to query an application for missing bits and pieces (mainly intermediate certs or OCSP responses).

Reply:
Thanks for providing the context on this. I agree this seems a deficiency in the API. Can you open an issue so we can track this independently of this PR? Just copying above text is sufficient.

It would be easy enough to allow an application to provide a `std::function<std::optional<X509_Certificate> (Some_Context)>` that is called when we are missing a certificate. And likewise OCSP responses. However what contextual information an application might need in order to find the desired data is unclear to me at this time.

Reply:
See: #3124